content/blog/fast-data-processing-pipeline-for-predicting-flight-delays-using-apache-apis-kafka-spark-streaming-and-machine-learning-part-3.md

```diff
@@ -0,0 +1,295 @@
+
+* Making input and output data available to independent consumers
+* Managing and evaluating multiple models and easily deploying new models
+* Monitoring and analyzing models, with historical and real-time data.
+
+![MapR Architectures](/img/mapr-architectures.png)
+
+Architectures for these types of applications are discussed in more detail in the eBooks: Machine Learning Logistics, Streaming Architecture, and Microservices and Containers.
+
+The following figure depicts the (simplified) data pipeline for this tutorial:
+
+* Flight trip data is published to a MapR Event Streams (ES) topic using the Kafka API. (Note this data contains the actual delay label, in the real world architecture the actual delay label would come later in a different topic, but to keep the tutorial code simple it is combined with the input data).
+* A Spark Streaming application subscribed to the first topic enriches the event with the flight predictions and publishes the results in JSON format to another topic. ( In the real world architecture there would be multiple consumers publishing model predictions, but to keep the tutorial code simple there is only one here).
+* A Spark Streaming application subscribed to the second topic stores the flight trip data and predictions in MapR Database using the Spark MapR Database Connector.
+* Apache Spark SQL, Apache Drill SQL, and Open JSON applications query MapR Database to analyze flight data and prediction performance.
+
+![MapR Database](/img/mapr-db.png)
+
+**How to Store the Data**
+
+One of the challenges, when you are processing lots of streaming data, is where do you want to store it? With a relational database and a normalized schema, related data is stored in different tables. Queries joining this data together can cause bottlenecks with lots of data. For this application, [MapR Database JSON](https://docs.datafabric.hpe.com/62/MapR-DB/developing_client_applications_for_mapr_db.html), a high-performance NoSQL database, was chosen for its scalability and flexible ease of use with JSON. MapR Database and a denormalized schema scale, because data that is read together is stored together.
+
+![Storage Model](/img/storage-model.png)
+
+With MapR Database (HBase API or JSON API, a table is automatically partitioned into tablets across a cluster by key range, providing for scalable and fast reads and writes by row key.
```

Comment:
Place an end-parenthesis after “JSON API”. Maintain the comma after the parenthesis.

![image](https://user-images.githubusercontent.com/77496261/137183229-e9865512-c03e-418c-b14b-4865b521fd5a.png)

Reply:
done